lib/Onyx.d.ts

```diff
@@ -0,0 +1,243 @@
+import {Keys, Values} from '.';
+import * as Logger from './Logger';
+
+type KeyList = keyof Keys;
+
+declare type ConnectOptions<K extends KeyList> = {
+    key: K;
+    statePropertyName?: string;
+    withOnyxInstance?: any; // TODO: Type this.
+    callback?: (value: Values[K] | undefined) => void;
+    initWithStoredValues?: boolean;
+    waitForCollectionCallback?: boolean;
+    selector?: any; // TODO: Type this.
+};
+
+declare type UpdateOperation<K extends KeyList> = {
+    onyxMethod: 'set' | 'merge' | 'mergeCollection';
+    key: K;
+    value: Values[K];
+};
+
+declare type InitOptions = {
+    keys?: Record<string, unknown>;
+    initialKeyStates?: Partial<Record<KeyList, Values[KeyList]>>;
```

Comment:
```suggestion
type OnyxKeysValues = {
    [Key in KeyList]: Values[Key];
};

declare type InitOptions = {
    keys?: Record<string, unknown>;
    initialKeyStates?: Partial<OnyxKeysValues>;
    safeEvictionKeys?: KeyList[];
    maxCachedKeysCount?: number;
    captureMetrics?: boolean;
    shouldSyncMultipleInstances?: boolean;
    debugSetState?: boolean;
};
```

Reply:
@blazejkustra  It works as long I don't add collection keys like `` `${ONYXKEYS.COLLECTION.DOWNLOAD}${'attachment1'}` ``, otherwise I start getting errors:

```ts
    Onyx.init({
        keys: ONYXKEYS,
        initialKeyStates: { // <- error
            [ONYXKEYS.ACCOUNT]: {id: 'id1'},
            [ONYXKEYS.IS_LOADING_PAYMENT_METHODS]: false,
            [`${ONYXKEYS.COLLECTION.DOWNLOAD}${'attachment1'}`]: {url: 'download_url'},
        },
        safeEvictionKeys: [ONYXKEYS.ACCOUNT],
    });
```

```
Type '{ [x: string]: false | { id: string; } | { url: string; }; account: { id: string; }; isLoadingPaymentMethods: false; }' is not assignable to type 'Partial<OnyxKeysValues>'.
  'string' and '`download_${string}`' index signatures are incompatible.
    Type 'false | { id: string; } | { url: string; }' is not assignable to type '{ url: string; } | undefined'.
      Type 'boolean' is not assignable to type '{ url: string; }'.
```

If I remove the collection key everything works as expected.

Reply:
> Typescript infers such literal strings as [x: string]: string; - which causes loose typings here -> [playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/MYewdgzgLgBAhjAvDA3gKBp+AuGByOPAGgywCNs8zjTMBtAAwBIU9g8BfBgXUvbQ4BuNFACeABwCmMAIJIYYqSABm8YQHp1WAHoB+IA)

In order to make it work we have to use `as const here` 😬  This should work 👇 

```ts
Onyx.init({
    keys: ONYXKEYS,
    initialKeyStates: { // <- error
        [ONYXKEYS.ACCOUNT]: {id: 'id1'},
        [ONYXKEYS.IS_LOADING_PAYMENT_METHODS]: false,
        [`${ONYXKEYS.COLLECTION.DOWNLOAD}${'attachment1'}` as const]: {url: 'download_url'},
    },
    safeEvictionKeys: [ONYXKEYS.ACCOUNT],
});
```